e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -0,0 +1,230 @@
+
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest/retry"
+	"kpt.dev/configsync/e2e/nomostest/testkubeclient"
+	"kpt.dev/configsync/e2e/nomostest/testlogger"
+	"kpt.dev/configsync/e2e/nomostest/testwatcher"
+	"kpt.dev/configsync/pkg/kinds"
+)
+
+// PortForwarder creates a port forwarding from localhost to a Deployment in the
+// cluster. Assumes the Deployment has a single Pod, and keeps the port forward
+// open if the Pod is deleted.
+type PortForwarder struct {
+	// kubeClient is used for list/get operations on k8s objects in the cluster
+	kubeClient *testkubeclient.KubeClient
+	// watcher is used to watch k8s objects in the cluster
+	watcher *testwatcher.Watcher
+	// logger is used for logging
+	logger *testlogger.TestLogger
+	// retryTimeout is used to determine the retry duration
+	retryTimeout time.Duration
+	// ctx should be a cancellable Context. Once ctx is cancelled, the port forward
+	// will be closed and will no longer be recreated.
+	ctx context.Context
+	// kubeConfigPath is the path to the kube config file for the cluster
+	kubeConfigPath string
+	// ns is the Namespace of the Deployment/Pod
+	ns string
+	// deployment is the name of the Deployment. Assumes a single Pod replica.
+	deployment string
+	// port is the port on the Pod to forward to
+	port string
+	// setPortCallback is a callback which will be invoked whenever the Pod changes and the
+	// port-forward is recreated.
+	setPortCallback func(int, string)
+	// mux is a mutex to synchronize getting/setting the port-forward value
+	mux sync.Mutex
+	// localPort is the local port which forwards to the Pod
+	localPort int
+	// pod is the name of the current Pod that is being forwarded to
+	pod string
+}
+
+// PortForwardOpt is an optional parameter for PortForwarder
+type PortForwardOpt func(pf *PortForwarder)
+
+// WithSetPortCallback registers a callback that will be invoked whenever the
+// deployment's underlying pod changes and the port-forward is recreated.
+func WithSetPortCallback(setFn func(int, string)) PortForwardOpt {
+	return func(pf *PortForwarder) {
+		pf.setPortCallback = setFn
+	}
+}
+
+// NewPortForwarder creates a new instance of a PortForwarder
+func NewPortForwarder(
+	ctx context.Context,
+	kubeClient *testkubeclient.KubeClient,
+	watcher *testwatcher.Watcher,
+	logger *testlogger.TestLogger,
+	retryTimeout time.Duration,
+	kubeConfigPath, ns, deployment, port string,
+	opts ...PortForwardOpt,
+) *PortForwarder {
+	pf := &PortForwarder{
+		kubeClient:     kubeClient,
+		watcher:        watcher,
+		logger:         logger,
+		ctx:            ctx,
+		kubeConfigPath: kubeConfigPath,
+		ns:             ns,
+		deployment:     deployment,
+		port:           port,
+		retryTimeout:   retryTimeout,
+	}
+	for _, opt := range opts {
+		opt(pf)
+	}
+	return pf
+}
+
+// GetPort returns the current localhost port which forwards to the pod.
+// lazily starts the port forward when called the first time.
+func (pf *PortForwarder) GetPort() (int, error) {
```

Comment:
nit... i have feelings about getters... feel free to take action or make a follow up PR or just tell me you dont care.

I think the go authors recommend avoiding "get" as a prefix for getters funcs. I don't know if I always agree with that tho. I tend to go with a get- prefix if it's a dumb function that just returns a value and avoid the get- prefix when it does some work to compute the result. This is a bit of a hybrid, due to the caching, but regardless, I think `LocalPort()` is a slightly more informative name. It also helps distinguish from the remote port being proxied to. Another option would be to have an `Address()` with the `localhost:` prefix but without a protocol. That would work even better if the name of the struct was `DeploymentPodProxy` or something, because it would be the "address of the proxy", which makes a little more sense than "address of the port forwarder". A port forwarder "forwards" as the verb, or "starts forwarding" but a proxy would have "start", "stop", and "address" as verbs (or just "address" if it automatically starts and stops).

Reply:
TLDR: "proxy" sounds like a service you start and stop. "port forwarder" sounds like a utility you have to actuate with "forward". So if we rename the struct to something like "proxy" then the verbs can be more service-like: start, stop, address.

Reply:
renamed to `LocalPort` and `start`

There are some existing interfaces which expect the port rather than an address (e.g. `GitProvider`), so I think at this point it makes sense to keep the port abstraction. I don't feel too strongly about the name of the struct/package - I think something like `DeploymentPodProxy` makes sense too